autorest/azure/async.go

```diff
@@ -0,0 +1,165 @@
+package azure
+
+import (
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/Azure/go-autorest/autorest"
+	"github.com/Azure/go-autorest/autorest/date"
+)
+
+const (
+	// HeaderAsyncOperation is the Azure header providing the URL from which to obtain the
+	// OperationResource for an asynchronous (aka long running) operation.
+	HeaderAsyncOperation = "Azure-AsyncOperation"
+)
+
+const (
+	// OperationCanceled says the underlying operation was canceled.
+	OperationCanceled string = "Canceled"
+	// OperationFailed says the underlying operation failed.
+	OperationFailed string = "Failed"
+	// OperationSucceeded says the underlying opertion succeeded.
+	OperationSucceeded string = "Succeeded"
+)
+
+// OperationError provides additional detail when an operation fails or is canceled.
+type OperationError struct {
+	// Code provides an invariant error code useful for troubleshooting, aggregration, and analysis.
+	Code string `json:"code,omitempty"`
+	// Message indicates what error occurred and what can be done to address the issue.
+	Message string `json:"message,omitempty"`
+}
+
+// Error implements the error interface returnin a string containing the code and message.
+func (oe OperationError) Error() string {
+	return fmt.Sprintf("Azure Operation Error -- code(%v), message(%v)", oe.Code, oe.Message)
```

Comment:
I suggest using %q for those. They'll 

a. add `"` quotes for you
b. Escape new lines etc properly so that you'll end up with a single line string. You can just say `Code=%q Message=%q`.

The communication language here with `--`s, `code(...)` etc seems a bit unusual to me. It's better to keep it simple:

    return fmt.Sprintf("Azure operation failed: Code=%q Message=%q")

is a ton more clearer and well formatted and consistent with rest of the error messages we have.